fix ZeroDivisionError in utils.bottleneck

egg-west opened the pull request:
**ZeroDivisionError** occurs when `cuda_prof_exec_time` is small enough.
This situation is normal for a project that has little CUDA work. 

Or someone does not make his work transferred to CUDA successfully. In this time he profiles the code, this error occurs.

egg-west committed:
fix ZeroDivisionError

ZeroDivisionError occurs when cuda_prof_exec_time is small enough.
This situation is normal for a project that has little CUDA work. 

Or someone does not make his work transferred to CUDA successfully. In this time he profiles the code, this error occurs.

soumith commented on torch/utils/bottleneck/__main__.py:
```diff
@@ -223,9 +223,10 @@ def main():
-    pct_diff = cuda_prof_exec_time - cpu_prof_exec_time / cuda_prof_exec_time
-    if abs(pct_diff) > 0.05:
-        print_autograd_prof_summary(autograd_prof_cpu, 'CPU', autograd_prof_sortby, autograd_prof_topk)
+    if cuda_prof_exec_time > 1e-6:
```
why not

```
if cuda_prof_exec_time > 0:
```

egg-west replied:
I did such a test. In python  `1.0/1e-8 = 100'000'000.0`
But `1.0/1e-9 = 999999999.9999999`. 
Then consider it use second, 1e-6 should be enough, or should it be 1e-8?

apaszke replied:
I'm confused by what your test means. If we're fixing a zero-division error then we should check for zeros

apaszke commented on torch/utils/bottleneck/__main__.py:
```diff
@@ -223,9 +223,10 @@ def main():
-    if abs(pct_diff) > 0.05:
-        print_autograd_prof_summary(autograd_prof_cpu, 'CPU', autograd_prof_sortby, autograd_prof_topk)
+    if cuda_prof_exec_time > 1e-6:
+        pct_diff = cuda_prof_exec_time - cpu_prof_exec_time / cuda_prof_exec_time
```
I'm wondering if this shouldn't have been
```python
(cuda_prof_exec_time - cpu_prof_exec_time) / cuda_prof_exec_time
```
cc @zou3519

egg-west replied:
I think you are right.

zou3519 replied:
Yes, that looks reasonable.

I'm not sure what the type of 'cuda_prof_exec_time' and 'cpu_prof_exec_time' are -- if they're python ints we should not be using the '/' operator (or we should `from future import __division__` at the top of this file so it works out.

egg-west replied:
cuda_prof_exec_time should be a float.

egg-west committed:
edit the bound and add a bracket

take the advice from advice and fix a newly discovered problem.

apaszke commented on torch/utils/bottleneck/__main__.py:
```diff
@@ -223,9 +223,10 @@ def main():
-    pct_diff = cuda_prof_exec_time - cpu_prof_exec_time / cuda_prof_exec_time
-    if abs(pct_diff) > 0.05:
-        print_autograd_prof_summary(autograd_prof_cpu, 'CPU', autograd_prof_sortby, autograd_prof_topk)
+    if cuda_prof_exec_time > 0.0:
```
Actually in which case can this be 0? I don't fully understand. You function *will take some time*, so it should be ok.

egg-west replied:
Actually, I was recurrent a case related to performance. I remember that has little cuda operation.
When I try to use bottleneck to profile it, this error occurs, so I start to fix it.

zou3519 replied:
@egg-west what do you mean by lack of bracket?

I'm able to reproduce this with the following things:
- empty script -- records nothing on the profiler
- `import torch; torch.empty([])` -- records nothing on the profiler

In general the CUDA time should always be nonzero because we're recording a CUDA event, running the function, and then recording another CUDA event, unless that function does nothing (and even then, python overhead adds some time). This is true even if the program doesn't use CUDA. The divide by zero error happens only when the profiler records nothing.

egg-west replied:
'lack of bracket' cause the priority error in `pct_diff = (cuda_prof_exec_time - cpu_prof_exec_time) / cuda_prof_exec_time`
I will test it again later.

zou3519 commented on torch/utils/bottleneck/__main__.py:
```diff
@@ -223,9 +223,10 @@ def main():
-    pct_diff = cuda_prof_exec_time - cpu_prof_exec_time / cuda_prof_exec_time
-    if abs(pct_diff) > 0.05:
-        print_autograd_prof_summary(autograd_prof_cpu, 'CPU', autograd_prof_sortby, autograd_prof_topk)
+    if cuda_prof_exec_time > 0.0:
```
@egg-west: could you please change this to instead check if the profiling result `autograd_prof_cpu` has no rows?

zou3519 reviewed (dismissed):
Please change the check to check if the profiling result is empty. Thank you for noticing the problem and submitting a fix, @egg-west!

egg-west committed:
change the check in line 225

Change the check of `cuda_prof_exec_time` to check the length of cpu profile events.

zou3519 reviewed (approved):
Please fix the lint, other than that, lgtm. Thanks @egg-west!

egg-west committed:
update __main__.py fix lint

line 230: remove space for blank line.

egg-west commented:
@apaszke PTAL.

zou3519 commented:
@egg-west could you rebase this on top of master so that the circle-ci tests run?

You can do that with:
```
git pull --rebase upstream master
git push -f
```

egg-west commented:
@zou3519  Thank you for all the help. I have done rebase, but there are still 2 test failed.
